### DISCUSSION TITLE
ezcox包问题求解

### DISCUSSION THREAD

**Discussion by zhaoliang0302** [ShixiangWang/self-study#88] [Q&A] [answered]
翔哥好，在使用你的ezcox包时有几个问题不太明白，还请帮忙解答下
- `ezcox`这个函数中的covariates和controls分别是什么意思，我的理解是covariates是想要研究的目标变量，control是可能对目标变量有影响的变量作为校正。我搜了一些cox回归的资料没搞明白
- `filter_ezcox`这个函数中的levels如何设定？我看你的[vignettes](https://cran.r-project.org/web/packages/ezcox/vignettes/ezcox.html)中设置成`c("0", "2")`，没搞明白含义是什么。还有`type`参数的contrast、ref和both代表什么？不清楚什么时候该用什么。我看`survival`包中的cox分析似乎没有这些参数设置
- `show_models`中合并model的意义是什么？
```r
# Merge all models and drop control variables
show_models(mds, merge_models = TRUE, drop_controls = TRUE)
```

这些好像都是统计知识，但是我自己检索了资料发现这些问题自己解决不了，还请翔哥有时间帮忙解答下，感谢！

**Comment by ShixiangWang** [comment, marked as answer]
@zhaoliang0302 这个包的初衷是批处理Cox模型的，所以之前开发的时候很多设定都是围绕这个展开的。

- 这里的协变量和控制变量本质上都是Cox模型中的协变量。我这个参数的名称设定可能会让人有些困惑，我之前也回答过类似的问题。我举例你就比较好理解了。

假设我们有3个基因A, B, C，还有患者的年龄信息。如果我要分别对基因建立Cox模型，同时将年龄信息纳入模型（校正年龄的影响）。那么：

```
covariats = c("A", "B", "C")
controls = "age"
```

会建立3个模型

Cox ~ A + age

Cox ~ B + age

Cox ~ C + age

在程序里面会对age打个标记，后续绘图可以隐藏这种我们可能不关注的变量。

2. filter_ezcox其实就是一个数据集的过滤行操作，这个完全你可以自己做。我就加了一些参数方便过滤一些特定的行，比如有患者的不同阶段的信息，例如临床阶段stage有I, II, III这3个因子，那么结果是每个因子都有结果的，我们可以只保留特定的，如`I`的结果。

3. `show_models`就是指Cox模型，例如第1点提到的3个基因的模型，那么`show_models()`可以在一个图中分别展示它们，为了简便，我们可能会需要隐藏不重要的信息，只保留基因（不展示校正的变量age），就是第3点的设定。

这些倒不怎么设计统计，你有兴趣可以翻翻源代码。

**Reply by zhaoliang0302** [reply]
多谢！
ezcox中的多因素cox都是引入control这个校正变量的，但在实际应用中我看到绝大多数文章没有设置control这样的校正变量，而比较多应用在这些场景：
- 多个基因A、B、C、D等，纳入这些基因表达量通过多因素cox构建风险模型，这些基因都作为协变量，没有校正变量的引入，这种在ezcox如何实现呢？
- 对基因X、年龄、WHO grade、突变等做多因素cox来说明基因X是否为独立预后因素，这些变量都是作为协变量而没有校正变量


在ezcox中：
```r
zz = ezcox(lung, covariates = c("sex", "ph.ecog"), controls = "age", return_models=TRUE)
mds = get_models(zz)
str(mds, max.level = 1)
#> List of 2
#>  $ Surv ~ sex + age    :List of 19
#>   ..- attr(*, "class")= chr "coxph"
#>   ..- attr(*, "Variable")= chr "sex"
#>  $ Surv ~ ph.ecog + age:List of 22
#>   ..- attr(*, "class")= chr "coxph"
#>   ..- attr(*, "Variable")= chr "ph.ecog"
#>  - attr(*, "class")= chr [1:2] "ezcox_models" "list"
#>  - attr(*, "has_control")= logi TRUE
```
上面代码中的模型会给出2个list，每个list分别计算sex和ph.ecog变量，那可以构建Cox ~ sex/ph.ecog这样的模型吗（同时纳入sex和ph.ecog变量）？

**Reply by ShixiangWang** [reply]
你说的这种场景并不是ezcox核心想做的事情，ezcox是为了方便大规模筛查。你这个其实就是建立一个多便利Cox模型，直接使用coxph就可以了。ezcox也是封装这个函数做的。

当然，你也可以用ezcox干这个事情：

```
library(survival)
library(ezcox)

ezcox(lung, covariates = "sex", controls = c("ph.ecog", "age"))
show_forest(lung, covariates = "sex", controls = c("ph.ecog", "age"), merge_models = TRUE)

```

本质上讲ezcox主要干这么一件事情，covariates有多少个就建多少个模型，而controls里面的变量都会被拼到模型中去。

```
Cox ~ sex (+ ph.ecog + age)
```

这里括号里面的都可以通过 controls 进行设定，有多少个变量，后面就会加多少个。

**Reply by zhaoliang0302** [reply]
我尝试了下，这样确实和`coxph`结果一样的，也就是说把covariates设置成一个，controls设置成多个，这种情况下等同于`coxph`函数中`surv(time, status) ~ age + sex + ph.ecog`这样的形式，而不再是controls代表校正变量？

**Reply by zhaoliang0302** [reply]
covariates一但设置不止一个变量最后的结果和coxph就不一样了

**Reply by ShixiangWang** [reply]
> 我尝试了下，这样确实和`coxph`结果一样的，也就是说把covariates设置成一个，controls设置成多个，这种情况下等同于`coxph`函数中`surv(time, status) ~ age + sex + ph.ecog`这样的形式，而不再是controls代表校正变量？

surv(time, status) ~ age + sex + ph.ecog 这个模型里面看你怎么看待变量，如果你想关注的是性别对生存的影响，那么其他两个就是被控制的（它们的影响被校正了），如果你关注的是age，那么同样推理。这里模型都是同一个，看你关注的变量。

最简单的，Cox ~ x1，我们研究x1的影响，但由于x2也可以对生存造成影响，那么Cox ~ x1 + x2就可以得到去除x2影响的x1结果。这里的结果主要是**x1的系数**。

**Reply by ShixiangWang** [reply]
> covariates一但设置不止一个变量最后的结果和coxph就不一样了

这个不可能的，我代码内部就是分开建模的。

**Reply by zhaoliang0302** [reply]
我搞明白了，ezcox中的covariates其实就是存放批量变量的，controls为空时就是对covariates中每个变量做单因素cox，controls不为空就表示分别对covariates中的每个变量和controls的变量结合后批量做多因素cox。我说的独立预后模型、风险模型其实就是covariates是一个变量，因为不需要批量做，最后的结果和coxph是一样的

**Reply by ShixiangWang** [reply]
对的。

**Reply by zhaoliang0302** [reply]
感谢翔哥！
